scripts/source-maintenance.sh

```diff
@@ -230,14 +238,24 @@ git grep "ifn?def .*_SQUID_" |
 #
 # Scan for file-specific actions
 #
-FORKPOINT=`git fork-point`
-if test "x$CHANGEDONLY" = "xyes" && test "x$FORKPOINT" != "x" ; then
-    FILES=`git diff -r $FORKPOINT --name-only`
-else
+FORKPOINT=""
+FILES=""
+if test "x$ONLYCHANGEDSINCE" = "x" ; then
     FILES=`git ls-files`
 fi
-echo "files: $FILES"
-exit 0
+if test "x$FILES" = "x" && test "x$ONLYCHANGEDSINCE" = "xfork" ; then
+    FORKPOINT=`git merge-base --fork-point upstream/master`
```

Comment:
The branch may not be forked off `master` and may not be using an `upstream` repository. If there is no way around hard-coding these two names, then please mention this assumption in the `--only-changed-since` documentation.

Reply:
This naming assumption is a killer change. It will make the script unusable for maintenance by the master server where maintenance happens.

Reply:
> Amos: This naming assumption is a killer change. It will make the script unusable for maintenance by the master server where maintenance happens.

I agree with Francesco. Yes, hard-coding this particular assumption is unfortunate, but it does not break anything and certainly does not make the script unusable anywhere. I cannot think of any proof for the two assertions quoted above. Do not like the hard-coded logic? Supply the right SHA or do not use the new command-line option at all! The option works for many but not all use cases, and that is OK.

The current hard-coded upstream/master default is the right default for the reasons specified by Francesco. I am not aware of any better choices.

The only improvement I can think of is adding logic to find the official branch the feature branch is actually based on (e.g., upstream/v4 instead of upstream/master), but I do not think it is worth doing.

FWIW, I have a trivial script that sets the environment variable called `$base` which I use in many commands. Usually, it is computed just like Francesco is computing it, but I can obviously overwrite it for special cases. I use that variable as a parameter to many git commands. It is very convenient in many use cases! I will use that variable as the new command line parameter value.

I see no valid reasons for such an outcome (other than lack of time and/or cooperation), but if we fail to convince Amos that this hard-coded default is acceptable, we can remove all support for the special `fork` value. I think it would be a step backwards because it is the best default (that we know of) for lots of use cases.

Reply:
> The branch may not be forked off `master` and may not be using an `upstream` repository. If there is no way around hard-coding these two names, then please mention this assumption in the `--only-changed-since` documentation.

Documented in the comment on top

Reply:
It would be best to keep command-line option documentation in one place -- printUsage(), but that is a subject of another change request.

Reply:
I'm talking about the fork point being hardcoded as upstream/master is the killer. If it were the default for a value passed in that would be fine.

The scripts deal with multiple repo upstream. Only one of them can be called upstream, the other has a different name.
Also, for our QA process to review changes before merge they are done on the maintenance-N branches instead of silently on backport or directly on master branch.

Reply:
> I'm talking about the fork point being hardcoded as upstream/master is the killer. If it were the default for a value passed in that would be fine.

I am not sure how to interpret the above correctly, but given your strong objection to something that has no known grounds for such an objection, I suspect there is still a basic misunderstanding of the effect/scope of the proposed hard-coded functionality. The proposed fork point calculation algorithm (which hard-codes `upstream/master`) is activated for one and only one value of the new optional command-line parameter called `--only-changed-since`:

1. The script user is free not to use `--only-changed-since` at all, preserving the current script functionality. The proposed optional feature cannot break anything because it does not change the default script behavior in any way.
2. The script user is free to use `--only-changed-since` with values other than `fork`, bypassing the fork point calculation algorithm but still getting all the benefit of the proposed optional feature.
3. If the script user elects to say `--only-changed-since fork`, then (and only then) will the fork point calculation kick in and use a hard-coded `upstream/master` value. Francesco has said that updated PR code documents this caveat. I will check that it does.

Implementation bugs notwithstanding, I see no "PR killing" functionality or flaws here. If you do, please describe exactly what will happen that should not happen.

> The scripts deal with multiple repo upstream. Only one of them can be called upstream, the other has a different name.

The fact that developer's (including automation's) works space may have many repositories seems to be irrelevant to the assertion that the proposed feature is a "PR killing" functionality. A user who does not like the hard-coded `upstream/master` base branch name will not use `--only-changed-since fork`.

> for our QA process to review changes before merge they are done on the maintenance-N branches instead of silently on backport or directly on master branch.

A developer can specify the correct forking point X by using `--only-changed-since X`. That X value (usually a commit SHA) can be based on any parent branch of developer choosing. The proposed functionality does not restrict X to `upstream/master` commits in any way.

Reply:
Doh. I see. Not sure what I was thinking before.